File: 2024/07/prog_17686_[3차]파일명정렬.py
def solution(files):
    answer = []
    
    # 1. 헤드 사전순 / 2. 숫자 순 => [(헤드, 숫자, 파일명)]
    sort_standard = []
    
    for file in files:
        # 넘버 인덱스 범위를 찾기 
        num_idx_start, num_idx_end = 0, len(file)-1 # 테일이 없어서 num_idx_end가 계산 안될 경우를 대비해서 초기화를 마지막 인덱스로 
        
        for i in range(len(file)):
            f = file[i]
            if f.isdigit():
                # 넘버 구간 시작 
                if num_idx_start == 0:
                    num_idx_start = i
            else:
                # 넘버 구간 끝
                if num_idx_start != 0:
                    num_idx_end = i-1
                    break
                    
        # 헤드 
        head = file[:num_idx_start]
        # 넘버 
        number = file[num_idx_start:num_idx_end+1]
        # 헤드를 소문자로 변환
        head = head.lower()
        # 숫자형으로 변환 
        number = int(number)
        
        sort_standard.append((head, number, file))
        
    # 정렬 
    sort_standard = sorted(sort_standard, key = lambda x: [x[0], x[1]])
    
    for h, n, f in sort_standard:
        answer.append(f)
    
    return answer

File: 2025/07/prog_17686_[3차]파일명정렬.js
function solution(files) {
    var answer = [];
    let total = []
    
    for (let i=0; i < files.length; i++) {
        const file = files[i].split("")
        const origin = files[i]
        // 헤드, 넘버, 테일 나누기 
        
        // 헤드 나누기
        let head = ""
        while (file.length > 0) {
            const f = file[0]
            // 숫자라면 
            if (f >= '0' && f <= '9') {
                break
            }
            head += f
            file.shift()
        }
                
        // 넘버 나누기 
        let number = ""
        while (file.length > 0 && number.length < 5) {
            const f = file[0] 
            // 숫자 -> false, 문자 -> true
            // 숫자가 아니라면 
            if (!(f >= '0' && f <= '9')) {
                break
            }
            number += f.trim()
            file.shift()
        }
        
        total.push([origin, head, number])
    }
    
    // 정렬 
    total.sort((a, b) => {
        // 헤드가 같다면 숫자순 정렬
        if (a[1].toLowerCase() === b[1].toLowerCase()) {
            return Number(a[2]) - Number(b[2]) 
        }
        // 헤드 정렬 
        else {
            return a[1].toLowerCase().localeCompare(b[1].toLowerCase())
        }
    })
    
    total.forEach((t) => answer.push(t[0]))
    return answer;
}